File: txt_to_srt.py
import re
import sys

def format_time(total_seconds):
    """Convert total seconds to SRT time format (HH:MM:SS,000)"""
    hours = total_seconds // 3600
    minutes = (total_seconds % 3600) // 60
    seconds = total_seconds % 60
    return f"{hours:02}:{minutes:02}:{seconds:02},000"

def parse_timestamp(timestamp):
    """Convert a timestamp (MM:SS or HH:MM:SS) into total seconds"""
    parts = list(map(int, timestamp.split(":")))
    if len(parts) == 2:  # Format MM:SS
        return parts[0] * 60 + parts[1]
    elif len(parts) == 3:  # Format HH:MM:SS
        return parts[0] * 3600 + parts[1] * 60 + parts[2]
    return 0

def youtube_to_srt(input_file, output_file):
    """Convert YouTube subtitle format to SRT format"""
    with open(input_file, "r", encoding="utf-8") as infile:
        lines = [line.strip() for line in infile.readlines()]

    re_time = re.compile(r"^\d{1,2}:\d{2}(:\d{2})?$")  # Matches MM:SS or HH:MM:SS
    srt_output = []
    subtitle_number = 1
    start_time, end_time = None, None
    subtitle_text = []

    for line in lines:
        if re_time.match(line):  # If it's a timestamp
            # Process previous subtitle if exists
            if start_time and subtitle_text:
                srt_output.append(f"{subtitle_number}")
                srt_output.append(f"{start_time} --> {end_time}")
                srt_output.append(" ".join(subtitle_text))
                srt_output.append("")
                subtitle_number += 1
                subtitle_text = []

            # Convert timestamp to total seconds
            start_seconds = parse_timestamp(line)
            start_time = format_time(start_seconds)
            end_time = format_time(start_seconds + 2)  # Default duration: 2s

        elif start_time:
            subtitle_text.append(line)

    # Process last subtitle if exists
    if start_time and subtitle_text:
        srt_output.append(f"{subtitle_number}")
        srt_output.append(f"{start_time} --> {end_time}")
        srt_output.append(" ".join(subtitle_text))
        srt_output.append("")

    # Write to output file
    with open(output_file, "w", encoding="utf-8") as outfile:
        outfile.write("\n".join(srt_output) + "\n")

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python youtube_to_srt.py <input_file> <output_file>")
        sys.exit(1)

    input_file = sys.argv[1]
    output_file = sys.argv[2]

    youtube_to_srt(input_file, output_file)
    print("Conversion completed successfully!")

File: txt_to_srt.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

// formatTime converts total seconds to SRT time format (HH:MM:SS,000)
func formatTime(totalSeconds int) string {
	hours := totalSeconds / 3600
	minutes := (totalSeconds % 3600) / 60
	seconds := totalSeconds % 60
	return fmt.Sprintf("%02d:%02d:%02d,000", hours, minutes, seconds)
}

// Converts a timestamp string (e.g., "1:00:05") to total seconds
func parseTimestamp(timestamp string) int {
	parts := strings.Split(timestamp, ":")
	if len(parts) == 2 { // Format MM:SS
		minutes, _ := strconv.Atoi(parts[0])
		seconds, _ := strconv.Atoi(parts[1])
		return minutes*60 + seconds
	} else if len(parts) == 3 { // Format HH:MM:SS
		hours, _ := strconv.Atoi(parts[0])
		minutes, _ := strconv.Atoi(parts[1])
		seconds, _ := strconv.Atoi(parts[2])
		return hours*3600 + minutes*60 + seconds
	}
	return 0
}

func youtubeToSrt(inputFile, outputFile string) error {
	file, err := os.Open(inputFile)
	if err != nil {
		return fmt.Errorf("error opening input file: %v", err)
	}
	defer file.Close()

	output, err := os.Create(outputFile)
	if err != nil {
		return fmt.Errorf("error creating output file: %v", err)
	}
	defer output.Close()

	scanner := bufio.NewScanner(file)
	reTime := regexp.MustCompile(`^\d{1,2}:\d{2}(:\d{2})?$`) // Matches MM:SS or HH:MM:SS
	var srtOutput []string
	subtitleNumber := 1
	var startTime, endTime string
	var subtitleText []string

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())

		// Check if the line is a timestamp
		if reTime.MatchString(line) {
			// Process previous subtitle if exists
			if startTime != "" && len(subtitleText) > 0 {
				srtOutput = append(srtOutput, fmt.Sprintf("%d", subtitleNumber))
				srtOutput = append(srtOutput, fmt.Sprintf("%s --> %s", startTime, endTime))
				srtOutput = append(srtOutput, strings.Join(subtitleText, " "))
				srtOutput = append(srtOutput, "")
				subtitleNumber++
				subtitleText = nil
			}

			// Convert timestamp to total seconds
			startSeconds := parseTimestamp(line)
			startTime = formatTime(startSeconds)

			// Default end time (2s after start)
			endTime = formatTime(startSeconds + 2)
		} else if startTime != "" {
			// Collect subtitle text
			subtitleText = append(subtitleText, line)
		}
	}

	// Process last subtitle if exists
	if startTime != "" && len(subtitleText) > 0 {
		srtOutput = append(srtOutput, fmt.Sprintf("%d", subtitleNumber))
		srtOutput = append(srtOutput, fmt.Sprintf("%s --> %s", startTime, endTime))
		srtOutput = append(srtOutput, strings.Join(subtitleText, " "))
		srtOutput = append(srtOutput, "")
	}

	// Write to output file
	writer := bufio.NewWriter(output)
	for _, line := range srtOutput {
		_, _ = writer.WriteString(line + "\n")
	}
	writer.Flush()

	return nil
}

func main() {
	if len(os.Args) != 3 {
		fmt.Println("Usage: go run youtube_to_srt.go <input_file> <output_file>")
		os.Exit(1)
	}

	inputFile := os.Args[1]
	outputFile := os.Args[2]

	err := youtubeToSrt(inputFile, outputFile)
	if err != nil {
		fmt.Println("Error:", err)
		os.Exit(1)
	}

	fmt.Println("Conversion completed successfully!")
}
